presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java

```diff
@@ -2064,6 +2114,84 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
         return analyzePropertyManager;
     }
 
+    @Override
+    public QualifiedObjectName redirectTable(Session session, QualifiedObjectName tableName, WarningCollector warningCollector)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(tableName, "tableName is null");
+        requireNonNull(warningCollector, "warningCollector is null");
+
+        Set<QualifiedObjectName> visitedTableNames = new HashSet<>();
+        visitedTableNames.add(tableName);
+        int count = TABLE_REDIRECTION_MAX_TIMES;
+        while (count-- > 0) {
+            Optional<QualifiedObjectName> redirectedTableName = Optional.empty();
+            Optional<CatalogMetadata> catalog = getOptionalCatalogMetadata(session, tableName.getCatalogName());
+            if (catalog.isPresent()) {
+                CatalogMetadata catalogMetadata = catalog.get();
+                CatalogName catalogName = catalogMetadata.getConnectorId(session, tableName);
+                ConnectorMetadata metadata = catalogMetadata.getMetadataFor(catalogName);
+                redirectedTableName = metadata.redirectTable(session.toConnectorSession(catalogName), tableName.asSchemaTableName())
+                        .map(name -> convertFromSchemaTableName(name.getCatalogName()).apply(name.getSchemaTableName()));
+            }
+            if (redirectedTableName.isEmpty()) {
+                return tableName;
+            }
+            warningCollector.add(new PrestoWarning(REDIRECTED_TABLE, format("Reference to table or view %s is redirected to %s", tableName, redirectedTableName.get())));
+            tableName = redirectedTableName.get();
+            if (!visitedTableNames.add(tableName)) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Table redirection forms a loop");
+            }
+        }
+        throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Table redirection happened more than %d times", TABLE_REDIRECTION_MAX_TIMES));
```

Comment:
If this exception is raised, administrator should have a way to understand what's wrong.
Consider including the redirects list in the message.